### ISSUE TITLE
CSE optimization in aggregation

### ISSUE THREAD

**Issue by zl862933** [pola-rs/polars#20305]
### Description

I've noticed that polars doesn't work with cse expressions for a very common class of cases, namely groupby then agg. However SQL based expression optimizers such as duckdb can do CSE optimization for this case quite simply.

I think this usage scenario is very common, especially when I have cse optimization when using .select directly on a single dataframe, however when used in groupby (like selecting on a group of dataframes at the same time) it doesn't and is very confusing.

```
import polars as pl

data = {"a": [1, 2], "b": [3, 4]}
df = pl.DataFrame(data).lazy()

print("===case1===")
print(
    df.select(pl.corr("a", "b").alias("expr1"), pl.corr("a", "b").alias("expr2")).explain(
        optimized=True
    )
)

print("===case2===")
print(
    df.group_by("a")
    .agg(pl.corr("a", "b").alias("expr1"), pl.corr("a", "b").alias("expr2"))
    .explain(optimized=True)
)
```

**Comment by borchero** [comment]
I do not think this is an issue. In your example, you run two separate queries -- no caching happens across queries. You can circumvent this with `collect_all`.

```python
lf = df.select(
    pl.corr("a", "b").alias("expr1"),
    pl.corr("a", "b").alias("expr2"),
)
print(pl.explain_all([lf, lf]))
```

prints

```
SINK_MULTIPLE
  PLAN 0:
    CACHE[id: 5ffcc0cd-54bc-462c-9202-d31c40a43000]
      simple π 2/2 ["expr1", "expr2"]
        SELECT [col("__POLARS_CSER_0xdae42fc4120318e8").alias("expr1"), col("__POLARS_CSER_0xdae42fc4120318e8").alias("expr2")]
           WITH_COLUMNS:
           [col("a").pearson_correlation([col("b")]).alias("__POLARS_CSER_0xdae42fc4120318e8")] 
            DF ["a", "b"]; PROJECT["a", "b"] 2/2 COLUMNS
  PLAN 1:
    CACHE[id: 5ffcc0cd-54bc-462c-9202-d31c40a43000]
      simple π 2/2 ["expr1", "expr2"]
        SELECT [col("__POLARS_CSER_0xdae42fc4120318e8").alias("expr1"), col("__POLARS_CSER_0xdae42fc4120318e8").alias("expr2")]
           WITH_COLUMNS:
           [col("a").pearson_correlation([col("b")]).alias("__POLARS_CSER_0xdae42fc4120318e8")] 
            DF ["a", "b"]; PROJECT["a", "b"] 2/2 COLUMNS
END SINK_MULTIPLE
```

**Comment by zl862933** [comment]
> I do not think this is an issue. In your example, you run two separate queries -- no caching happens across queries. You can circumvent this with `collect_all`.
> 
> lf = df.select(
>     pl.corr("a", "b").alias("expr1"),
>     pl.corr("a", "b").alias("expr2"),
> )
> print(pl.explain_all([lf, lf]))
> prints
> 
> ```
> SINK_MULTIPLE
>   PLAN 0:
>     CACHE[id: 5ffcc0cd-54bc-462c-9202-d31c40a43000]
>       simple π 2/2 ["expr1", "expr2"]
>         SELECT [col("__POLARS_CSER_0xdae42fc4120318e8").alias("expr1"), col("__POLARS_CSER_0xdae42fc4120318e8").alias("expr2")]
>            WITH_COLUMNS:
>            [col("a").pearson_correlation([col("b")]).alias("__POLARS_CSER_0xdae42fc4120318e8")] 
>             DF ["a", "b"]; PROJECT["a", "b"] 2/2 COLUMNS
>   PLAN 1:
>     CACHE[id: 5ffcc0cd-54bc-462c-9202-d31c40a43000]
>       simple π 2/2 ["expr1", "expr2"]
>         SELECT [col("__POLARS_CSER_0xdae42fc4120318e8").alias("expr1"), col("__POLARS_CSER_0xdae42fc4120318e8").alias("expr2")]
>            WITH_COLUMNS:
>            [col("a").pearson_correlation([col("b")]).alias("__POLARS_CSER_0xdae42fc4120318e8")] 
>             DF ["a", "b"]; PROJECT["a", "b"] 2/2 COLUMNS
> END SINK_MULTIPLE
> ```

I don't think you understand what I'm saying at all.

In my example, case1 and case2 are obviously two separate queries. And they are two completely different queries. The first doesn't use group_by, and the second use group_by.

I don't understand what your example is trying to illustrate.

**Comment by borchero** [comment]
Ahh, sorry @zl862933, I misunderstood this issue, never mind then 😅 I can reproduce what you meant 👀
